# varibles and Data types

# Exercise 1.1: Memory Exploration

1. Create variables of different types (int, float, string, list, tuple) 
2. Print the ID (memory address) of each variable using the id() function

In [3]:


# int
print('int')
ammount = 100
dozens = 2
print("type of variables ammount and dozens ",type(ammount),type(dozens))
print('memory address of ammount is ', id(ammount))
print('memory address of dozens is ', id(dozens))
print('\n')

# float
print('float')
weight = 56.70
height = 154.6
print("type of variables weight and height ",type(weight),type(height))
print('memory address of weight is ', id(weight))
print('memory address of height is ', id(height))
print('\n')

# string
print('string')
name = 'pavan'
location = 'Hyderabad'
print("type of variables name and location ",type(name),type(location))
print('memory address of name is ', id(name))
print('memory address of location is', id(location))
print('\n')


# list : These are Homogenous
print('list')
l = [1, 2, 'pavan', 18]
print("type of l is ",type(l))
print('meomry address of list l is ', id(l))
print('\n')

# tuple
print('tuple')
t = (1, 4, 'hello')
print("type of t is ", type(t))
print('memory address of tuple t is ', id(t))



int
type of variables ammount and dozens  <class 'int'> <class 'int'>
memory address of ammount is  4304131536
memory address of dozens is  4303939920


float
type of variables weight and height  <class 'float'> <class 'float'>
memory address of weight is  4345907888
memory address of height is  4343973872


string
type of variables name and location  <class 'str'> <class 'str'>
memory address of name is  4330964464
memory address of location is 4346193520


list
type of l is  <class 'list'>
meomry address of list l is  4345960896


tuple
type of t is  <class 'tuple'>
memory address of tuple t is  4345958656


3. Create a second variable that references the same object as one of your first variables.
4. Modify the original variable and observe what happens to the second one (for both
mutable and immutable types).

In [7]:
first_variable = 10
second_variable = first_variable

print(f'first_variable is {first_variable} and address is {id(first_variable)} ')
print(f'second_variable is {second_variable} and address is {id(second_variable)} \n')

print("After modyfying first_variable:")
first_variable = 20
print(f'first_variable is {first_variable} and address is {id(first_variable)} ')
print(f'second_variable is {second_variable} and address is {id(second_variable)} \n')

first_variable is 10 and address is 4303940176 
second_variable is 10 and address is 4303940176 

After modyfying first_variable:
first_variable is 20 and address is 4303940496 
second_variable is 10 and address is 4303940176 



5. Include comments explaining the behavior you observe.

comments are not executable code

# Exercise 1.2: Variable Scope Investigation

Create a function that demonstrates variable scope in Python:
1. Define global variables outside the function.
2. Define local variables inside the function with the same names.
3. Try to modify a global variable both with and without the global keyword.
4. Print the IDs of all variables before and after modifications.
5. Explain what happens and why in your comments.

In [14]:

def summ(a : int, b : int) -> int:

    # declaring local variables inside the function
    a = 3
    b = 15
    print(f'inside function values of a : {a} and b : {b}')
    s = a + b
    print(f'After execution of function inside summ is ',s)
    return s


# declaring global variables outside the function
a = 5
b = 4
print(f'Before modifying the global variables values of a : {a} and  b : {b}')
print(f'Now global variables are sending into a function ', summ(a, b))
print(f'After function Execution values of global variables are a : {a} and b : {b} and their sum is {summ(a,b)}')
print(f'global sum of {a}and {b} is ', a+ b)

# global variables are not change inside of a function




Before modifying the global variables values of a : 5 and  b : 4
inside function values of a : 3 and b : 15
After execution of function inside summ is  18
Now global variables are sending into a function  18
inside function values of a : 3 and b : 15
After execution of function inside summ is  18
After function Execution values of global variables are a : 5 and b : 4 and their sum is 18
global sum of 5and 4 is  9


# Part 2: Data Types and Type Conversion


# Exercise 2.1: Type Exploration

Create a program that:
1. Creates at least one variable of each of these types: int, float, complex, bool, str, and
None.
2. Uses the type() function to verify the type of each variable.
3. Uses isinstance() to check if variables are of specific types.
4. Demonstrates at least three examples where Python automatically converts types in
expressions.
5. Includes comments documenting your observations about type behavior.
Exercise 2.2: Type Conversion Challenge
Write a function that:
1. Takes a string input containing a mixture of numbers and text (e.g., "I am 25 years old
and my height is 5.9 feet")
2. Extracts all numbers from the string and converts them to their appropriate numeric types
(int or float).
3. Returns a tuple containing two lists: one with all integers found and one with all floats
found.
4. Handles potential conversion errors gracefully

In [25]:
# 1
list_of_different_variables = [18, 'pavan', 5.9, 3+4j, True, None]
print(list_of_different_variables)

for var in list_of_different_variables:
    print(f'type of variable {var}',type(var))




[18, 'pavan', 5.9, (3+4j), True, None]
type of variable 18 <class 'int'>
type of variable pavan <class 'str'>
type of variable 5.9 <class 'float'>
type of variable (3+4j) <class 'complex'>
type of variable True <class 'bool'>
type of variable None <class 'NoneType'>


In [18]:
# 2. Using the isinstance() function to check the type of each variable in the list

for var in list_of_different_variables:
    if isinstance(var, int):
        print(f'{var} is an integer')
    elif isinstance(var, float):
        print(f'{var} is a float')
    elif isinstance(var, str):
        print(f'{var} is a string')
    elif(isinstance(var, complex)):
        print(f'{var} is a complex number')
    elif isinstance(var, bool):
        print(f'{var} is a boolean')
    elif var is None:
        print(f'{var} is NoneType')
        

18 is an integer
pavan is a string
5.9 is a float
(3+4j) is a complex number
True is an integer
None is NoneType


In [22]:
# 4. Automatically converting the python variables from one type to another type
# 5. Explain with comments
# Example 1:

x = 10  # int   
y = 3.14 # float

z = x + y # Automatic type conversion (int to float)

print(f'Automatic type conversion example: {x} + {y} = {z} (type of z is {type(z)})')

# Example 2:

x = 3
s = 'Hi! '

print(x*s) # Automatic type conversion (int to str)


# example 3:
x = 5
y = True
z = x + y # Automatically convert True to 1 (int)
print(z)



Automatic type conversion example: 10 + 3.14 = 13.14 (type of z is <class 'float'>)
Hi! Hi! Hi! 
6


In [28]:
# 6. Takes a string input from the user containing mixture of numbers and text
s = input("Enter a string containing mixture of numbers and text: ")
print("You entered the string:", s)

You entered the string: hello my name pavan and i am 22 years old and my weight is 65.7


In [29]:
# Extract all the numbers from the string and convert them into their data types

import re

numbers = re.findall(r'\d+', s)  # Find all sequences of digits
converted_numbers = [float(num) if '.' in num else int(num) for num in numbers]

print("Extracted numbers and their types:")
for num in converted_numbers:
    print(f'{num} is of type {type(num)}')


Extracted numbers and their types:
22 is of type <class 'int'>
65 is of type <class 'int'>
7 is of type <class 'int'>


In [35]:
# 7. return with tuples of lists containing with integers and float values in a string

import re

def extract_numbers(text: str):
    # Correct regex: Match float numbers first, then integers
    numbers = re.findall(r'\d+\.\d+|\d+', text)

    integers = []
    floats = []

    for num in numbers:
        try:
            if '.' in num:
                floats.append(float(num))
            else:
                integers.append(int(num))
        except ValueError:
            # Handle any unexpected conversion error
            continue

    return (integers, floats)

# Test string
s = "The price is 100 dollars and the weight is 56.7 kg."
integers, floats = extract_numbers(s)

print("Integers extracted from the string:", integers)
print("Floats extracted from the string:", floats)


Integers extracted from the string: [100]
Floats extracted from the string: [56.7]


# Part 3: Number Systems and Representation

Exercise 3.1: Base Converter
Create a set of functions that:
1. Converts decimal integers to binary, octal, and hexadecimal strings without using built-in
functions like bin(), oct(), or hex().
2. Converts binary, octal, and hexadecimal strings to decimal integers without using int(x,
base).
3. Demonstrates your functions with at least five different numbers.
4. Verifies your results by comparing with Python's built-in conversion functions.
